File: iteration2/Models/Student.java
package iteration2.Models;

import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.logging.FileHandler;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

public class Student extends Person {

    private String id;
    private int registerDate;
    private Boolean isGraduate;
    private int semesterNo;
    private Transcript transcript;

    private ArrayList<String> errors;
    private Advisor advisor;
    private HashMap<Course, Boolean> selectedCourses;

    private static String FILE_PATH="/iteration2/Logs/Student.log";
    private static Logger logger=Logger.getLogger(Student.class.getName());
    private static FileHandler fileHandler;

    static {
        try {
            fileHandler = new FileHandler(System.getProperty("user.dir")+FILE_PATH,true);
            logger.addHandler(fileHandler);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void customLog(boolean type,String message){
        SimpleFormatter formatter =new SimpleFormatter();
        fileHandler.setFormatter(formatter);
        if(type)
            logger.info(message);
        else
            logger.warning(message);
    }
    public Student(String name, String surname, String ssn, Character gender, String id, Boolean isGraduate, int registerDate, int semesterNo, Transcript transcript, Advisor advisor) {
        super(name, surname, ssn, gender);

        this.id = id;
        this.registerDate = registerDate;
        this.isGraduate = isGraduate;
        this.semesterNo = semesterNo;
        this.transcript = transcript;
        this.advisor = advisor;
        this.errors = new ArrayList<>();

        customLog(true, "Student constructor method is called and new student object is generated.");
    }

    public String getId() {
        return id;
    }

    public int getRegisterDate() {
        return registerDate;
    }

    public Boolean getIsGraduate() {
        return isGraduate;
    }

    public int getSemesterNo() {
        return semesterNo;
    }

    public Transcript getTranscript() {
        return transcript;
    }

    public ArrayList<String> getErrors() {
        return errors;
    }

    public Advisor getAdvisor() {
        return advisor;
    }

    public HashMap<Course, Boolean> getSelectedCourses() {
        return selectedCourses;
    }

    public void setSemesterNo(int semesterNo) {
        this.semesterNo = semesterNo;
    }

    public void setTranscript(Transcript transcript) {
        this.transcript = transcript;
    }

    public void setSelectedCourses(HashMap<Course, Boolean> selectedCourses) {
        this.selectedCourses = selectedCourses;
    }

    public void addError(String errorMessage) {
        this.errors.add(errorMessage);
    }

    @Override
    public String getName() {
        return super.getName();
    }

    @Override
    public String getSurname() {
        return super.getSurname();
    }

    @Override
    public String getSsn() {
        return super.getSsn();
    }

    @Override
    public Character getGender() {
        return super.getGender();
    }

    @Override
    public JSONObject toJson() {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("name", getName());
        jsonObject.put("surname", getSurname());
        jsonObject.put("ssn", getSsn());
        jsonObject.put("gender", getGender());
        jsonObject.put("id", getId());
        jsonObject.put("isGraduate", getIsGraduate());
        jsonObject.put("registerDate", getRegisterDate());
        jsonObject.put("semesterNo", getSemesterNo());
        jsonObject.put("Transcript", getTranscript().toJson());
        JSONObject advisorjson = new JSONObject();
        advisorjson.put("name", advisor.getName());
        advisorjson.put("surname", advisor.getSurname());
        advisorjson.put("ssn", advisor.getSsn());
        jsonObject.put("Advisor", advisorjson);
        jsonObject.put("errors", getErrors());

        return jsonObject;
    }


    public void addToSelectedCourses(Course course, Boolean status) {
        if (!selectedCourses.containsKey(course)) {
            selectedCourses.put(course, status);
            customLog(true, "Checked if the course is completed before and it is not completed. Added to selected courses");
        }
        else {
            customLog(false, "Checked if the course is completed before and it is completed. Course cannot be added");
        }
    }
}

File: iteration3/Models/Student.py
import json
import logging

from iteration3.Models.Person import Person
from abc import ABC, abstractmethod


class Student(Person):
    logger = logging.getLogger(__name__)
    file_handler = logging.FileHandler("RegistrationError.log")
    formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
    file_handler.setFormatter(formatter)
    logger.addHandler(file_handler)

    def __init__(self, name, surname, ssn, gender, id, isGraduate, registerDate, semesterNo, transcript, advisor):
        super().__init__(name, surname, ssn, gender)
        self.__id = id
        self.__registerDate = registerDate
        self.__isGraduate = isGraduate
        self.__semesterNo = semesterNo
        self.__transcript = transcript
        self.__advisor = advisor
        self.__errors = []
        self.__selectedCourses = []
        """self.custom_log(True, "Student constructor method is called and new student object is generated.")"""
    def custom_log(self, is_error, message):
        if is_error:
            self.logger.info(message)
        else:
            self.logger.warning(message)

    @property
    def id(self):
        return self.__id

    @property
    def register_date(self):
        return self.__registerDate

    @property
    def is_graduate(self):
        return self.__isGraduate

    @property
    def semester_no(self):
        return self.__semesterNo

    @property
    def transcript(self):
        return self.__transcript

    @property
    def errors(self):
        return self.__errors

    @property
    def advisor(self):
        return self.__advisor

    @property
    def selected_courses(self):
        return self.__selectedCourses

    @semester_no.setter
    def set_semester_no(self, semesterNo):
        self.__semesterNo = semesterNo

    @transcript.setter
    def set_transcript(self, transcript):
        self.__transcript = transcript

    @selected_courses.setter
    def set_selected_courses(self, selectedCourses):
        self.__selectedCourses = selectedCourses

    def add_error(self, error_message):
        self.errors.append(error_message)

    def to_json(self):
        json_object = {
            "name": self.name,
            "surname": self.surname,
            "ssn": self.ssn,
            "gender": self.gender,
            "id": self.id,
            "isGraduate": self.is_graduate,
            "registerDate": self.register_date,
            "semesterNo": self.semester_no,
            #"Transcript": self.transcript.to_json,
            "errors": self.errors,
            "Advisor": {
                "name": self.advisor.name,
                "surname": self.advisor.surname,
                "ssn": self.advisor.ssn
            }
        }
        return json_object

    def add_to_selected_courses(self, course, status):
        if course not in self.selected_courses:
            self.selected_courses.append([course, status])
            self.custom_log(True,
                            "Checked if the course is completed before and it is not completed. Added to selected courses")
        else:
            self.custom_log(False,
                            "Checked if the course is completed before and it is completed. Course cannot be added")
